{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9489ef50-6a85-4dd0-81cb-83d430445527",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "import math\n",
    "import csv\n",
    "import multiprocessing as multiprocessing\n",
    "from Bio import SeqIO\n",
    "from Bio import AlignIO\n",
    "import os\n",
    "import re\n",
    "from scipy import stats\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment \n",
    "from Bio.Seq import Seq \n",
    "from multiprocessing import Pool\n",
    "from FunctionsAnalyzeNanopore import ProcessFile\n",
    "from FunctionsAnalyzeNanopore import ConcatenateDataFrame\n",
    "from FunctionsAnalyzeNanopore import CalculateMismatchIndels\n",
    "from FunctionsAnalyzeNanopore import CountSelectedMismatches\n",
    "from FunctionsAnalyzeNanopore import CountAllMismatches\n",
    "\n",
    "sns.set_style(\"whitegrid\")\n",
    "\n",
    "num_cores = multiprocessing.cpu_count()\n",
    "# num_cores = 6\n",
    "pool_obj = multiprocessing.Pool()\n",
    "\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "plt.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8d2eb272-9df3-4ff2-8b9d-b1ef6ee59de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#SET EXPERIMENT TO ANALYZE. save this in some file eventually\n",
    "\n",
    "GFPseqF = \"tgcttcagccgctaccccgaccacatgaagcagcacgacttcttcaagtccgccatgcccgaaggctacgtccaggagcgcaccatcttcttcaaggacgacggcaactacaagacccgcgccgaggtgaagttcgagggcgacaccctggtgaaccgcatcgagctgaagggcatcgacttcaaggaggacggcaacatcctggggcacaagctggagtacaactacaacagccacaacgtctatatcatggccgacaagcagaagaacggcatcaa\"\n",
    "eveMSE_BFP_F = \"GCAGCACGAGGTGGCAGTGGCCAGATACTGCGACCTCCCTAGCAAACTGGGGCACAAGCTTAATTGAtctagagcatgTAATACGACTCACTATAGCATGTAATACGACTCACTATAGCATGcaggttacccggtactgcataacaatggaacccgaaccgtaactgggacagatcgaaaagctggcctggtttctcgctgtgtgtgccgtgttaatccgtttgccatcagcgagattattagtcaattgcagttgcagcgtttcgctttcgtcctcgtttcactttcgagttagactttattgcagcatcttgaacaatcgtcgcagtttggtaacacgctgtgccatactttcatttagacggaatcgagggaccctggactataatcgcacaacgagaccgggttgcgaagtcagggcattccgccgatctagccatcgccatcttctgcgggcgtttgtttgtttgtttgctgggattagccaagggcttgacttggaatccaatcccgatccctagcccgatcccaatcccaatcccaatcccttgtccttttcattagaaagtcataaaaacacataataatgatgtcgaagggattaggggcgaattcggccggccgcttcgagc\"\n",
    "eveMSE_BFP_F_polym = \"GCAGCACGAGGTGGCAGTGGCCAGATACTGCGACCTCCCTAGCAAACTGGGGCACAAGCTTAATTGAtctagagcatgTAATACGACTCACTATAGCATGTAATACGACTCACTATAGCATGcaggttacccggtactgcataacaatggaacccgaaccgtaactgggacagatcgaaaagctggcctggtttctcgctgtgtgtgccgtgttaatccgtttgccatcagcgagattattagtcaattgcagttgcagcgtttcgctttcgtcctcgtttcactttcgagttagactttattgcagcatcttgaacaatcgtcgcagtttggtaacacgctgtgccatactttcatttagacggaatcgagggaccctggactataatcgcacaacgagaccgggttgcgaagtcagggcattccgccgatctagccatcgccatcttctgcgggcgtttgtttgtttgtttgctgggattagccaagggcttgacttggaatccaatccTgatccctagcccgatcccaatcccaatcccaatcccttgtccttttcattagaaagtcataaaaacacataataatgatgtcgaagggattaggggcgaattcggccggccgcttcgagc\"\n",
    "eveMSE_end_F=\"tgccaatgcaaattgcgggcgcaatataacccaataatttgaagtaactggcaggagcgaggtatccttcctggttacccggtactgcataacaatggaacccgaaccgtaactgggacagatcgaaaagctggcctggtttctcgctgtgtgtgccgtgttaatccgtttgccatcagcgagattattagtcaattgcagttgcagcgtttcgctttcgtcctcgtttcactttcgagttagactttattgcagcatcttgaacaatcgtcgcagtttggtaacacgctgtgccatactttcatttagacggaatcgagggaccctggactataatcgcacaacgagaccgggttgcgaagtcagggcattccgccgatctagccatcgccatcttctgcgggcgtttgtttgtttgtttgctgggattagccaagggcttgacttggaatccaatcccgatccctagcccgatcccaatcccaatcccaatcccttgtccttttcattagaaagtcataaaaacacataataatgatgtcgaagggattaggggcgcgcaggtccaggcaacgcaat\"\n",
    "GFPseqF = eveMSE_BFP_F_polym\n",
    "#GFPseqF = GFPseqF.upper()\n",
    "#GFPseqR = str(Seq(GFPseqF).reverse_complement())\n",
    "\n",
    "IsDuplex = 0\n",
    "Bidirectional = 0\n",
    "# Dir = \"/Volumes/Elements/Nanopore/20230124_94_MEJFS002_AK/basecalling-high-accuracy/pass/\"\n",
    "# DirOut = \"/Volumes/Elements/Nanopore/20230124_94_MEJFS002_AK/plots/\"\n",
    "# Dir = \"/Volumes/Elements/Nanopore/20230124_94_MEJFS002_AK/guppyout_sup/pass/\"\n",
    "# DirOut = \"/Volumes/Elements/Nanopore/20230124_94_MEJFS002_AK/plots-sup/\"\n",
    "\n",
    "# Dir = \"/Volumes/Elements/Nanopore/20230124_104_MEJFS002_JFS/basecalling-260bps-high-accuracy/pass/\"\n",
    "# DirOut = \"/Volumes/Elements/Nanopore/20230124_104_MEJFS002_JFS/plots/\"\n",
    "CutStart = 15 #15 for gfp\n",
    "CutEnd = 75 #75 for gfp\n",
    "#Path = '/Volumes/Elements/Nanopore/20230124_104_controlGFP_JFS/'; IsDuplex = 1\n",
    "Path = '/Volumes/Elements/Nanopore/20221003_1253_MN36507_FAU75974_0ec2a181/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_4.csv' #redone\n",
    "# Path = '/Volumes/Elements/Nanopore/20230124_104_MEJFS002_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_4.csv' #redone\n",
    "# Path = '/Volumes/Elements/Nanopore/20230221_104_MEJFS003_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_4.csv'\n",
    "# # # Path = '/Volumes/Elements/Nanopore/20230328_104_MEJFS004_AK/'\n",
    "# Path = '/Volumes/Elements/Nanopore/20230803_94_MEJFS005_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_4.csv' #redone\n",
    "# #Path = '/Volumes/Elements/Nanopore/20230803_94_MEJFS006_AK/'\n",
    "# Path = '/Volumes/Elements/Nanopore/20230824_94_MEJFS007_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_4.csv' #redone\n",
    "# #Path = '/Volumes/Elements/Nanopore/20230905_94_MEYDT001_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_24.csv'\n",
    "# Path = '/Volumes/Elements/Nanopore/20231012_104_MEJFS008_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_4.csv' #redone\n",
    "# Path = '/Volumes/Elements/Nanopore/20231106_104_MEYDT002_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_24.csv' #redone\n",
    "# Path = '/Volumes/Elements/Nanopore/20231116_104_MEYDT003_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_8-45.csv' #redone\n",
    "# Path = '/Volumes/Elements/Nanopore/20231121_104_MEYDT004_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_4.csv' #redone\n",
    "# Path = '/Volumes/Elements/Nanopore/20231122_104_MEYDT005_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_1_23.csv' #redone\n",
    "# Path = '/Volumes/Elements/Nanopore/20231122_104_MEYDT006_AK/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_1_20.csv'\n",
    "# Path = '/Volumes/Elements/Nanopore/20240402_104_MEJFS010_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_eve2_BFP_T7.csv'; IsDuplex = 1; Bidirectional=1 ; CutStart = 124; CutEnd = 24\n",
    "# Path = '/Volumes/Elements/Nanopore/20240402_104_MEJFS011_AK/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_4.csv'; IsDuplex = 1; Bidirectional=0 ;CutStart = 15; CutEnd = 75\n",
    "# #Path = '/Volumes/Elements/Nanopore/20240403_104_MEYDT011_AK/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_eve2_end_31_20.csv'; Bidirectional=1\n",
    "# #Path = '/Volumes/Elements/Nanopore/20240403_104_MEYDT011_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_eve2_end_31_20.csv'; Bidirectional=1\n",
    "# Path = '/Volumes/Elements/Nanopore/20240408_104_MEYDT011_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_eve2_end_31_20.csv';IsDuplex = 1; Bidirectional=1; CutStart = 72; CutEnd = 23\n",
    "# Path = '/Volumes/Elements/Nanopore/20240529_104_MEJFS012_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_eve2_BFP_T7.csv'; IsDuplex = 0; Bidirectional=1 ; CutStart = 124; CutEnd = 24\n",
    "# Path = '/Users/julia//20240529_104_MEJFS012_JFS/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs.csv'; IsDuplex = 0; Bidirectional=1 ; CutStart = 124; CutEnd = 24\n",
    "Path = '/Volumes/Elements/Nanopore/20240610_94_MEYDT013_AK/' ; IsDuplex = 0 ; gRNAFile = 'gRNAs_eve2_BFP_T7.csv'; IsDuplex = 0; Bidirectional=1 ; CutStart = 70; CutEnd = 20\n",
    "\n",
    "\n",
    "# import barcodes for multiplexing\n",
    "MPxFile = Path + 'multiplexing.csv'\n",
    "gRNAFile = Path + gRNAFile\n",
    "\n",
    "if not IsDuplex:\n",
    "    Dir =  Path + 'guppyout_sup/pass/'\n",
    "    DirOut =  Path + 'plots-sup/'\n",
    "else:\n",
    "    Dir =  Path + 'doradoout_sup_duplex/'\n",
    "    DirOut =  Path + 'plots-sup-duplex/'\n",
    "    \n",
    "os.makedirs(DirOut,exist_ok = True)\n",
    "\n",
    "P7seq = 'CAAGCAGAAGACGGCATACGAGAT'\n",
    "R2seq = 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATC'\n",
    "P5seq = 'AATGATACGGCGACCACCGAGATC'\n",
    "R1seq = 'ACACTCTTTCCCTACACGACGCTCTTCCGATCT'\n",
    "IndexNEB = 'NNNNNN'\n",
    "PrimerFWD = P5seq + 'T' + R1seq\n",
    "PrimerREV = P7seq + IndexNEB + R2seq\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "29c29d19-faba-488b-b127-58858323562f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "range(0, 3)\n",
      "[0]\n"
     ]
    }
   ],
   "source": [
    "DictTranslate = str.maketrans({\"N\": \"0\",\"A\": \"1\",\"C\": \"2\",\"T\": \"3\",\"G\": \"4\",\"-\": \"5\",\"I\": None})\n",
    "DictTranslateREV = str.maketrans({ \"0\":\"N\", \"1\":\"A\", \"2\":\"C\", \"3\":\"T\", \"4\":\"G\", \"5\":\"-\"})\n",
    "DictComp = str.maketrans({\"N\": \"N\",\"A\": \"T\",\"C\": \"G\",\"T\": \"A\",\"G\": \"C\",\"-\": \"-\",\"I\": \"I\"})\n",
    "\n",
    "if Bidirectional == 0:\n",
    "    GFPseqF = PrimerFWD+GFPseqF.upper()+str(Seq(PrimerREV).reverse_complement())\n",
    "    LibraryLength = len(GFPseqF)\n",
    "else:\n",
    "    GFPseqSF = ['NaN']*2\n",
    "    GFPseqSF[0] = PrimerFWD+GFPseqF.upper()+str(Seq(PrimerREV).reverse_complement())\n",
    "    GFPseqSF[1] = PrimerFWD+str(Seq(GFPseqF).reverse_complement()).upper()+str(Seq(PrimerREV).reverse_complement())\n",
    "    LibraryLength = len(GFPseqSF[0])\n",
    "    GFPseqF = GFPseqSF\n",
    "\n",
    "\n",
    "#print(GFPseqF)\n",
    "\n",
    "DictMultiplexing = {}\n",
    "ControlSamples = []\n",
    "index = 0\n",
    "with open(MPxFile, newline='') as csvfile:\n",
    "    reader = csv.reader(csvfile, dialect='excel')\n",
    "    for row in reader:\n",
    "        Sample = row[0]\n",
    "        Index = row[1]\n",
    "        Description = row[2]\n",
    "        Experiment = row[3]\n",
    "        try:\n",
    "            Genotype = row[4]\n",
    "            Type = row[5]\n",
    "            Time = row[6]\n",
    "        except:\n",
    "            Genotype = 'NaN'\n",
    "            Type = 'NaN'\n",
    "            Time = 0\n",
    "        DictMultiplexing[Sample] = {'Sample':Sample,'Index':Index,'Description':Description,'Experiment':Experiment,'Genotype':Genotype,'Type':Type,'Time':Time}\n",
    "        if Type == 'control' or Type == 'ctr':\n",
    "            ControlSamples.append(index)\n",
    "        index += 1     \n",
    "ExperimentSamples = range(0,len(DictMultiplexing))\n",
    "\n",
    "print(ExperimentSamples)\n",
    "print(ControlSamples)\n",
    "\n",
    "\n",
    "Files = [f for f in os.listdir(Dir) if re.match(r\"\\w.*\\.fastq\\.gz$\", f)]\n",
    "#print(\"\\n\".join(Files))\n",
    "#print(GFPseqF)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c56dd86-a10a-4518-90f6-e9ad907355d6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#RUN QUALITY THRESHOLDING, ALIGNMENT AND INDEX MATCHING\n",
    "FilesSelected = Files # SELECT FILES TO RUN\n",
    "#delete files that have already been run. comment to overwrite\n",
    "FilesDF = [f for f in os.listdir(DirOut) if re.match(r\"\\w.*DF\\.csv$\", f)]\n",
    "FilesSelectedCopy = FilesSelected.copy()\n",
    "for f in FilesSelected:\n",
    "    if f+'_DF.csv' in FilesDF:\n",
    "         FilesSelectedCopy.remove(f)\n",
    "FilesSelected = FilesSelectedCopy.copy()     \n",
    "print(len(FilesSelected))\n",
    "print(FilesSelected)\n",
    "\n",
    "Dictionaries = [DictTranslate,DictTranslateREV,DictComp,DictMultiplexing]\n",
    "MaxLength = 3000\n",
    "QualityThres = 10 #min average quality to keep read\n",
    "AlignmentThres = 1000 #alignment score threshold. #600 for GFP?\n",
    "Parameters = [MaxLength,LibraryLength,QualityThres,AlignmentThres,PrimerFWD,PrimerREV]\n",
    "AlignmentParameters = [2,0,-3,0] #match, mismatch, gap opening, gap extension (used to be -3)\n",
    "IndexPositions = [LibraryLength-len(P7seq)-len(IndexNEB),LibraryLength-len(P7seq)] #positions where indexes for multiplexing should be\n",
    "\n",
    "\n",
    "args = [(file,GFPseqF,Dir,DirOut,Dictionaries,Parameters,AlignmentParameters,IndexPositions) for file in FilesSelected]\n",
    "pool_obj.starmap(ProcessFile,args)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bea5e7a3-a056-499b-9954-491bdee8c45d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c48c7711-2a8b-43fe-b083-7456e853a46d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REF not found\n",
      "Unassigned not found\n"
     ]
    }
   ],
   "source": [
    "# COMBINE ALL CSV FILES TOGETHER\n",
    "# needs to have fun second cell to get DirOut\n",
    "\n",
    "FilesDF = [f for f in os.listdir(DirOut) if re.match(r\"^\\w.*DF\\.csv$\", f)]\n",
    "FilesIn = [f for f in os.listdir(DirOut) if re.match(r\"^\\w.*DFInsertions\\.csv$\", f)]\n",
    "FilesDF = sorted(FilesDF)\n",
    "FilesIn = sorted(FilesIn)\n",
    "# PropFiles = round(len(Files)*0.002)\n",
    "# FilesDF = FilesDF[0:PropFiles]\n",
    "# FilesIn = FilesIn[0:PropFiles]\n",
    "\n",
    "num_cores = multiprocessing.cpu_count()\n",
    "pool_obj = multiprocessing.Pool()\n",
    "args = [(FileSet,DirOut) for FileSet in [FilesDF,FilesIn]]\n",
    "[DFall,DFInall] = pool_obj.starmap(ConcatenateDataFrame,args)\n",
    "\n",
    "\n",
    "DFall['Sample'] = DFall['Sample'].astype(str)\n",
    "DFInall['Sample'] = DFInall['Sample'].astype(str)\n",
    "\n",
    "DFall = DFall.sort_values('Sample', ascending=True, key=lambda col: col.str.lower())\n",
    "DFInall = DFInall.sort_values('Sample', ascending=True, key=lambda col: col.str.lower())\n",
    "\n",
    "Experiments = DFInall['Sample'].unique()\n",
    "Descriptions = DFInall['Sample'].unique()\n",
    "i=0\n",
    "for sample in DFInall['Sample'].unique():\n",
    "    try:\n",
    "        Experiments[i] = DictMultiplexing[sample]['Experiment']\n",
    "        Descriptions[i] = DictMultiplexing[sample]['Description']\n",
    "    except:\n",
    "        print(sample+' not found')\n",
    "    i+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ad20649b-7e09-4982-81eb-bac9211e9135",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "# CALCULATE MISMATCHES TO REF AND INDELS\n",
    "num_cores = 4\n",
    "pool_obj = multiprocessing.Pool()\n",
    "\n",
    "MatrixMM = np.zeros((len(DFall['Sample'].unique()),LibraryLength),float)\n",
    "MatrixIndels = np.zeros((len(DFall['Sample'].unique()),LibraryLength),float)\n",
    "MatrixAllDiff = np.zeros((len(DFall['Sample'].unique()),LibraryLength),float)\n",
    "Ns = 0*len(DFall['Sample'].unique())\n",
    "REF = DFall[DFall['Sample'] == 'REF']\n",
    "REFgRNAs = REF.copy()\n",
    "REFgRNAsFWD = REF.copy()*0\n",
    "REFgRNAsREV = REF.copy()*0\n",
    "REFgRNAsCutSites = REF.copy()*0\n",
    "\n",
    "gRNAsDF = pd.read_csv(gRNAFile, sep = ',')\n",
    "for index, row in gRNAsDF.iterrows():\n",
    "    Start = row['Start']+len(PrimerFWD)\n",
    "    End = row['End']+len(PrimerFWD)\n",
    "    Orientation = row['Orientation']\n",
    "    REFgRNAs.iloc[:,Start-1:End] = 0\n",
    "    if Orientation == '+':\n",
    "        REFgRNAsFWD.iloc[:,Start-1:End] = 1\n",
    "        REFgRNAsCutSites.iloc[:,End-3] = 1\n",
    "    else:\n",
    "        REFgRNAsREV.iloc[:,Start-1:End] = 1\n",
    "        REFgRNAsCutSites.iloc[:,Start+2] = 1\n",
    "\n",
    "DFall[DFall['Sample'] == 'REF'] = REFgRNAs\n",
    "args = [(DFall,DFInall,sample,LibraryLength,REF) for sample in DFall['Sample'].unique()]\n",
    "[Mismatch, Indels, AllDiff,DFMutations, N, Samples] = zip(*pool_obj.starmap(CalculateMismatchIndels,args))\n",
    "print('done')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c2758639-19d0-4604-8c4a-15f6419982df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no update needed\n"
     ]
    }
   ],
   "source": [
    "#only when everything was run with the wrong order, create new multiplexing with reordered sample numbers\n",
    "try:\n",
    "    MPxFile2 = Path + 'multiplexing-reordered.csv'\n",
    "    DictMultiplexing2 = {}\n",
    "    ControlSamples2 = []\n",
    "    NewOrder = []\n",
    "    index = 0\n",
    "    with open(MPxFile2, newline='') as csvfile:\n",
    "        reader = csv.reader(csvfile, dialect='excel')\n",
    "        for row in reader:\n",
    "            Sample = row[0]\n",
    "            Index = row[1]\n",
    "            Description = row[2]\n",
    "            Experiment = row[3]\n",
    "            try:\n",
    "#                 Genotype = row[4]\n",
    "                Type = row[5]\n",
    "                Time = row[6]\n",
    "            except:\n",
    "                Genotype = 'NaN'\n",
    "                Type = 'NaN'\n",
    "                Time = 0\n",
    "            DictMultiplexing2[Sample] = {'Sample':Sample,'Index':Index,'Description':Description,'Experiment':Experiment,'Genotype':Genotype,'Type':Type,'Time':Time}\n",
    "            NewOrder.append(int(Sample))\n",
    "            if Type == 'control' or Type == 'ctr':\n",
    "                ControlSamples2.append(int(Sample)-1)\n",
    "            index += 1 \n",
    "    NewOrder.append(max(NewOrder)+1)\n",
    "    NewOrder.append(max(NewOrder)+1)\n",
    "    ExperimentSamples2 = range(0,len(DictMultiplexing2))\n",
    "    print(ExperimentSamples2)\n",
    "    print(ControlSamples2)\n",
    "    print(NewOrder)\n",
    "    Mismatch_temp = [0]*len(N)\n",
    "    DFMutations_temp = [0]*len(N)\n",
    "    Indels_temp =[0]*len(N)\n",
    "    N_temp = [0]*len(N)\n",
    "    Samples_temp = [0]*len(N)\n",
    "    Descriptions_temp = [0]*len(N)\n",
    "    count = 0\n",
    "    for i in NewOrder:\n",
    "        Mismatch_temp[i-1] = Mismatch[count]\n",
    "        DFMutations_temp[i-1] = DFMutations[count]\n",
    "        Indels_temp[i-1] = Indels[count]\n",
    "        N_temp[i-1] = N[count]\n",
    "        Samples_temp[i-1] = Samples[count]\n",
    "        Descriptions_temp[i-1] = Descriptions[count]\n",
    "        count += 1\n",
    "    print(Samples_temp)\n",
    "    Mismatch = Mismatch_temp\n",
    "    DFMutations = DFMutations_temp\n",
    "    Indels = Indels_temp\n",
    "    N = N_temp\n",
    "    Samples = Samples_temp\n",
    "    Descriptions = Descriptions_temp\n",
    "    ControlSamples = ControlSamples2\n",
    "    ExperimentSamples = ExperimentSamples2\n",
    "    print(Samples)\n",
    "except:\n",
    "    print('no update needed')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07bc51cb-5857-4dc2-8221-8eb12c590181",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23db75bc-c96e-43f9-8b6c-f2ddbe16152b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "347f3f6b-7c6a-4061-9271-82732e2c382a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# PLOT PROFILES MISMATCH/INDELS AND HEATMAPS AVERAGE SAMPLES\n",
    "\n",
    "Threshold = 5\n",
    "samplei = 0\n",
    "numcol=3\n",
    "fig1, axes1 = plt.subplots(math.ceil(len(Samples)/numcol), numcol, sharex=True, sharey=True, figsize=(20,10))\n",
    "fig2, axes2 = plt.subplots(math.ceil(len(Samples)/numcol), numcol, sharex=True, sharey=True, figsize=(20,10))\n",
    "for sample in DFall['Sample'].unique():\n",
    "    MM = Mismatch[samplei] \n",
    "    ID = Indels[samplei]\n",
    "    Neach = N[samplei]\n",
    "    plot = sns.lineplot(ax=axes1[math.ceil((samplei+1)/numcol)-1,samplei-((math.ceil((samplei+1)/numcol))-1)*numcol],x=range(0,len(MM)),y=MM*100)\n",
    "    plot.set(title=Samples[samplei]+' n='+str(Neach)+'\\n'+Descriptions[samplei],ylim=(0, Threshold),ylabel = '% mismatched', xlabel = 'position (bp)')\n",
    "    plot = sns.lineplot(ax=axes2[math.ceil((samplei+1)/numcol)-1,samplei-((math.ceil((samplei+1)/numcol))-1)*numcol],x=range(0,len(ID)),y=ID*100)\n",
    "    plot.set(title=Samples[samplei]+' n='+str(Neach)+'\\n'+Descriptions[samplei],ylim=(0, Threshold),ylabel = '% indel', xlabel = 'position (bp)')\n",
    "    MatrixMM[samplei,:] = MM\n",
    "    MatrixIndels[samplei,:] = ID\n",
    "    MatrixAllDiff[samplei,:] = AllDiff[samplei]\n",
    "    samplei += 1\n",
    "fig1.suptitle('% mismatched reads to GFP in each position from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "fig1.tight_layout()\n",
    "fig1.savefig(DirOut+\"allcombined_mismatched_samples_\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")\n",
    "fig1.clf()\n",
    "plt.close()\n",
    "fig2.suptitle('% indels in each position from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "fig2.tight_layout()\n",
    "fig2.savefig(DirOut+\"allcombined_indels_samples_\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")\n",
    "fig2.clf()\n",
    "plt.close()\n",
    "\n",
    "np.savetxt(DirOut+\"allcombined_MatrixMM.\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.csv\", MatrixMM, delimiter=\",\")\n",
    "np.savetxt(DirOut+\"allcombined_MatrixIndels.\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.csv\", MatrixIndels, delimiter=\",\")\n",
    "np.savetxt(DirOut+\"allcombined_Ns.\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.csv\",N,delimiter=\",\")\n",
    "\n",
    "Experiments = pd.DataFrame(Experiments)\n",
    "lut = dict(zip(set(Experiments[0]), sns.color_palette(\"viridis\",len(Experiments[0]))))\n",
    "row_colors = Experiments[0].map(lut).to_numpy()\n",
    "\n",
    "hmM = sns.clustermap(MatrixMM*100,col_cluster=False,row_cluster=False,figsize=(10, 10),vmin=0,vmax = Threshold, yticklabels=Descriptions, row_colors = row_colors, cbar_kws={'label': '% mismatch'})\n",
    "for Control in ControlSamples:\n",
    "    hmM.ax_heatmap.axhline(Control, color='white', lw=2)\n",
    "hmM.ax_heatmap.text(-15, 6,Experiments[0][0],rotation=90,color='w')\n",
    "#hmM.ax_heatmap.text(-15, 15,Experiments[0][8],rotation=90,color='w')\n",
    "#hmM.ax_heatmap.text(-15, 25,Experiments[0][20],rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(-15, 28.5,'R',rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(-15, 29.5,'U',rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(163, 28.75,'gRNAs',rotation=0,color='k')\n",
    "hmM.ax_cbar.set_position((0.075,0.6, .03, .2))\n",
    "hmM.fig.suptitle('% mismatched reads to GFP in each position from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "hmM.savefig(DirOut+\"allcombined_heatmapMismatched\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.png\",dpi=300)\n",
    "hmM.fig.clf()\n",
    "plt.close()\n",
    "\n",
    "hmM = sns.clustermap(MatrixIndels*100,col_cluster=False,row_cluster=False,figsize=(10, 10),vmin=0,vmax = Threshold, yticklabels=Descriptions, row_colors = row_colors, cbar_kws={'label': '% indels'})\n",
    "for Control in ControlSamples:\n",
    "    hmM.ax_heatmap.axhline(Control, color='white', lw=2)\n",
    "hmM.ax_heatmap.text(-15, 6,Experiments[0][0],rotation=90,color='w')\n",
    "#hmM.ax_heatmap.text(-15, 15,Experiments[0][8],rotation=90,color='w')\n",
    "#hmM.ax_heatmap.text(-15, 25,Experiments[0][20],rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(-15, 28.5,'R',rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(-15, 29.5,'U',rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(163, 28.75,'gRNAs',rotation=0,color='k')\n",
    "hmM.ax_cbar.set_position((0.075,0.6, .03, .2))\n",
    "hmM.fig.suptitle('% indels in each position from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "hmM.savefig(DirOut+\"allcombined_heatmapIndels\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.png\",dpi=300)\n",
    "hmM.fig.clf()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7d10bfe0-ee0e-4f51-99b7-5c9b668c1c81",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# SUBTRACT BACKGROUND FROM CONTROLS, REPLOT LINEPLOTS AND HEATMAPS\n",
    "\n",
    "# Threshold = 40\n",
    "\n",
    "fig1, axes1 = plt.subplots(math.ceil(len(Samples)/numcol), numcol, sharex=True, sharey=True, figsize=(20,10))\n",
    "fig2, axes2 = plt.subplots(math.ceil(len(Samples)/numcol), numcol, sharex=True, sharey=True, figsize=(20,10))\n",
    "\n",
    "# Mismatch_BG = Mismatch.copy()\n",
    "# Indels_BG = Indels.copy()\n",
    "MatrixMM_BG = np.zeros((len(DFall['Sample'].unique()),LibraryLength),float)\n",
    "MatrixIndels_BG = np.zeros((len(DFall['Sample'].unique()),LibraryLength),float)\n",
    "iREF = Samples.index('REF')\n",
    "MatrixMM_BG[iREF,:] = MatrixMM[iREF,:]\n",
    "MatrixIndels_BG[iREF,:] = MatrixMM[iREF,:]\n",
    "\n",
    "\n",
    "#\n",
    "#CtrMM = np.sum([a*b for a,b in zip(Mismatch,N)],0)/sum(N)\n",
    "#CtrID = np.sum([a*b for a,b in zip(Indels,N)],0)/sum(N)\n",
    "\n",
    "samplei=0\n",
    "for Exp in ExperimentSamples:\n",
    "    for Control in ControlSamples:\n",
    "        if Exp >= Control:\n",
    "            CtrMM = Mismatch[Control]\n",
    "            CtrID = Indels[Control]\n",
    "    MM = Mismatch[Exp] \n",
    "    ID = Indels[Exp]\n",
    "    MatrixMM_BG[Exp,:] = MM-CtrMM\n",
    "    MatrixIndels_BG[Exp,:] = ID-CtrID\n",
    "    plot = sns.lineplot(ax=axes1[math.ceil((samplei+1)/numcol)-1,samplei-((math.ceil((samplei+1)/numcol))-1)*numcol],x=range(0,len(MM)),y=(MM-CtrMM)*100)\n",
    "    plot.set(title=Samples[Exp]+' n='+str(N[Exp])+'\\n'+Descriptions[Exp],ylim=(0, Threshold),ylabel = '% mismatched', xlabel = 'position (bp)')\n",
    "    plot = sns.lineplot(ax=axes2[math.ceil((samplei+1)/numcol)-1,samplei-((math.ceil((samplei+1)/numcol))-1)*numcol],x=range(0,len(ID)),y=(ID-CtrID)*100)\n",
    "    plot.set(title=Samples[Exp]+' n='+str(N[Exp])+'\\n'+Descriptions[Exp],ylim=(0, Threshold),ylabel = '% indel', xlabel = 'position (bp)')\n",
    "    samplei+=1\n",
    "\n",
    "fig1.suptitle('% mismatched reads to GFP in each position after background substraction from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "fig1.tight_layout()\n",
    "fig1.savefig(DirOut+\"allcombined_mismatched_samples_-background_\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")\n",
    "fig1.clf()\n",
    "plt.close()\n",
    "fig2.suptitle('% indels in each position after background substraction from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "fig2.tight_layout()\n",
    "fig2.savefig(DirOut+\"allcombined_indels_samples_-background_\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")\n",
    "fig2.clf()\n",
    "plt.close()\n",
    "\n",
    "\n",
    "# REPLOT HEATMAPS AFTER BACKGROUND SUBTRACTION\n",
    "\n",
    "Experiments = pd.DataFrame(Experiments)\n",
    "lut = dict(zip(set(Experiments[0]), sns.color_palette(\"viridis\",len(Experiments[0]))))\n",
    "row_colors = Experiments[0].map(lut).to_numpy()\n",
    "\n",
    "hmM = sns.clustermap(MatrixMM_BG[:,len(PrimerFWD):LibraryLength-len(PrimerREV)]*100,col_cluster=False,row_cluster=False,figsize=(10, 10),vmin=0,vmax = Threshold, yticklabels=Descriptions, row_colors = row_colors, cbar_kws={'label': '% mismatch'},rasterized=True,xticklabels=50)\n",
    "for Control in ControlSamples:\n",
    "    hmM.ax_heatmap.axhline(Control, color='white', lw=1)\n",
    "    hmM.ax_heatmap.axhline(Control+1, color='white', lw=1)\n",
    "\n",
    "hmM.ax_heatmap.text(-15, 6,Experiments[0][0],rotation=90,color='w')\n",
    "#hmM.ax_heatmap.text(-15, 15,Experiments[0][8],rotation=90,color='w')\n",
    "#hmM.ax_heatmap.text(-15, 25,Experiments[0][20],rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(-15, 28.5,'R',rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(-15, 29.5,'U',rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(163, 28.75,'gRNAs',rotation=0,color='k')\n",
    "hmM.ax_cbar.set_position((0.075,0.6, .03, .2))\n",
    "hmM.fig.suptitle('% mismatched reads to GFP in each position after background subtraction from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "hmM.savefig(DirOut+\"allcombined_heatmapMismatched_noBG\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\",dpi=600)\n",
    "hmM.fig.clf()\n",
    "plt.close()\n",
    "\n",
    "hmM = sns.clustermap(MatrixIndels_BG[:,len(PrimerFWD):LibraryLength-len(PrimerREV)]*100,col_cluster=False,row_cluster=False,figsize=(10, 10),vmin=0,vmax = Threshold, yticklabels=Descriptions, row_colors = row_colors, cbar_kws={'label': '% indels'},rasterized=True,xticklabels=50)\n",
    "for Control in ControlSamples:\n",
    "    hmM.ax_heatmap.axhline(Control, color='white', lw=1)\n",
    "    hmM.ax_heatmap.axhline(Control+1, color='white', lw=1)\n",
    "hmM.ax_heatmap.text(-15, 6,Experiments[0][0],rotation=90,color='w')\n",
    "#hmM.ax_heatmap.text(-15, 15,Experiments[0][8],rotation=90,color='w')\n",
    "#hmM.ax_heatmap.text(-15, 25,Experiments[0][20],rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(-15, 28.5,'R',rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(-15, 29.5,'U',rotation=90,color='w')\n",
    "hmM.ax_heatmap.text(163, 28.75,'gRNAs',rotation=0,color='k')\n",
    "hmM.ax_cbar.set_position((0.075,0.6, .03, .2))\n",
    "hmM.fig.suptitle('% indels in each position after background subtraction from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "hmM.savefig(DirOut+\"allcombined_heatmapIndels_noBG\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\",dpi=600)\n",
    "hmM.fig.clf()\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "43ef3302-0496-4b61-8612-0a59b085401e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAATACGACTCACTATAG\n",
      "TAATA*GA*T*A*TATAG\n",
      "AGGTTACCCGGTACTGCATAACAATGGAACCCGAACCGTAACTGGGACAGATCGAAAAGCTGGCCTGGTTTCTCGCTGTGTGTGCCGTGTTAATCCGTTTGCCATCAGCGAGATTATTAGTCAATTGCAGTTGCAGCGTTTCGCTTTCGTCCTCGTTTCACTTTCGAGTTAGACTTTATTGCAGCATCTTGAACAATCGTCGCAGTTTGGTAACACGCTGTGCCATACTTTCATTTAGACGGAATCGAGGGACCCTGGACTATAATCGCACAACGAGACCGGGTTGCGAAGTCAGGGCATTCCGCCGATCTAGCCATCGCCATCTTCTGCGGGCGTTTGTTTGTTTGTTTGCTGGGATTAGCCAAGGGCTTGACTTGGAATCCAATCCTGATCCCTAGCCCGATCCCAATCCCAATCCCAATCCCTTGTCCTTTTCATTAGAAAGTCATAAAAACACATAATAATGATGTCGAAGGGATTAGGGG\n"
     ]
    }
   ],
   "source": [
    "#SELECT MUTATIONS\n",
    "\n",
    "#Which = Samples.index('E5')\n",
    "#Which = Samples.index('#15')\n",
    "#Which = Samples.index('6_4')\n",
    "\n",
    "gRNAsDF = pd.read_csv(gRNAFile, sep = ',')\n",
    "# count number of possible mutation on gRNAs\n",
    "mutation_limit=0\n",
    "if len(GFPseqF)>2:\n",
    "    GFPforCounting=GFPseqF\n",
    "else:\n",
    "    GFPforCounting=GFPseqF[0]\n",
    "    \n",
    "for index, row in gRNAsDF.iterrows():   \n",
    "    Start = row['Start']+len(PrimerFWD)\n",
    "    End = row['End']+len(PrimerFWD)\n",
    "    Orientation = row['Orientation']\n",
    "    gRNAseq = GFPforCounting[Start-1:End]\n",
    "    print(gRNAseq)\n",
    "    gRNAseqList = list(gRNAseq)\n",
    "    x=0 \n",
    "    while x < len(gRNAseqList): \n",
    "        if Orientation == '+'and gRNAseqList[x]==\"C\":\n",
    "                mutation_limit+=1\n",
    "                gRNAseqList[x]='*'\n",
    "        if Orientation == '-' and gRNAseqList[x]==\"G\":\n",
    "                mutation_limit+=1\n",
    "                gRNAseqList[x]='*'\n",
    "        x=x+1\n",
    "    GFPforCounting = GFPforCounting.replace(gRNAseq,''.join(gRNAseqList))\n",
    "\n",
    "isMutable = [i for i, ltr in enumerate(GFPforCounting) if ltr == '*']\n",
    "MutationsSelected = [0]*LibraryLength\n",
    "for i in isMutable:\n",
    "    MutationsSelected[i] = 1\n",
    "MutationsSelected = np.array(MutationsSelected) > 0\n",
    "MutationsSelected = np.transpose(MutationsSelected)\n",
    "# WhichControl = 1\n",
    "#MutationsSelected2 = (Mismatch[Which] - Mismatch[ControlSamples[0]]) > 0.005\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "412e26d1-91f7-4177-b440-c461f8c30e9e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# PLOT BAR PLOT W ALL MISMATCHES AND EACH TYPE BEFORE AND AFTER BG SUBSTRACTION. \n",
    "\n",
    "# fig1, axes1 = plt.subplots(3, 6, sharex=True, sharey=True, figsize=(60,10))\n",
    "# fig2, axes2 = plt.subplots(3, 6, sharex=True, sharey=True, figsize=(20,10))\n",
    "# fig3, axes3 = plt.subplots(3, 6, sharex=True, sharey=True, figsize=(20,10))\n",
    "# fig4, axes4 = plt.subplots(3, 6, sharex=True, sharey=True, figsize=(20,10))\n",
    "\n",
    "gRNAsboo = np.where(REFgRNAs.iloc[0,len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd]==0)\n",
    "gRNAsbooFWD = np.where(REFgRNAsFWD.iloc[0,len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd]==1)\n",
    "gRNAsbooREV = np.where(REFgRNAsREV.iloc[0,len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd]==1)\n",
    "gRNAsbooCuts = np.where(REFgRNAsCutSites.iloc[0,len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd]==1)\n",
    "\n",
    "Size = 500\n",
    "DFMutations_BG = list(DFMutations).copy()\n",
    "# CutStart = 0; CutEnd = 0\n",
    "#EXP1 act5C 1week\n",
    "# CtrMM1 = np.average([Mismatch[0],Mismatch[1],Mismatch[2],Mismatch[3],Mismatch[5]],0)\n",
    "# CtrMMeach1 = np.average([DFMutations[0]/N[0],DFMutations[1]/N[1],DFMutations[2]/N[2],DFMutations[3]/N[3],DFMutations[5]/N[5]],0)\n",
    "# CtrID1 = np.average([Indels[0],Indels[1],Indels[2],Indels[3],Indels[5]],0)\n",
    "# CtrMM1 = np.sum([a*b for a,b in zip(Mismatch,N)],0)/sum(N)\n",
    "# CtrID1 = np.sum([a*b for a,b in zip(Indels,N)],0)/sum(N)\n",
    "# CtrMMeach1 = np.sum([a*b for a,b in zip(DFMutations,N)],0)/sum(N)\n",
    "\n",
    "# CtrMM1 = Mismatch[0]\n",
    "# CtrMMeach1 = DFMutations[0]/N[0]\n",
    "# CtrID1 = Indels[0]\n",
    "if len(GFPseqF)==2:\n",
    "    GFPseqF=GFPseqF[0]\n",
    "\n",
    "GFPList = list(GFPseqF)[len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd]\n",
    "for Exp in range(0,len(Samples)):\n",
    "    for Control in ControlSamples:\n",
    "        if Exp >= Control:\n",
    "            CtrMM = Mismatch[Control]\n",
    "            CtrMMeach = DFMutations[Control]/N[Control]\n",
    "            CtrID = Indels[Control]\n",
    "    f, ax = plt.subplots(4,1,figsize = (25,15))\n",
    "    MM = Mismatch[Exp] \n",
    "    DFMutationsEach = DFMutations[Exp]/N[Exp]*100\n",
    "    DFMutations_BG[Exp] = DFMutations[Exp]/N[Exp]-CtrMMeach\n",
    "\n",
    "    DFMM = pd.DataFrame((MM)*100)\n",
    "    DFMM.iloc[len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd,:].plot(ax = ax[0], kind='bar', width = 1)\n",
    "    ax[0].set(title = Samples[Exp]+' n='+str(N[Exp])+'   '+Descriptions[Exp]+' total mismatches', ylim = [-5,Threshold],xlabel = 'position (bp)', ylabel = '% mismatch')\n",
    "    ax[0].set_xticklabels(GFPList, rotation = 0)\n",
    "    ax[0].grid(False)\n",
    "    ax[0].scatter(gRNAsbooFWD[0],[-3]*len(gRNAsbooFWD[0]), marker='4', cmap='summer', c = 'gray', s=Size, alpha = 0.75)\n",
    "    ax[0].scatter(gRNAsbooREV[0],[-3]*len(gRNAsbooREV[0]), marker='3', cmap='summer', c = 'gray', s=Size, alpha = 0.75)\n",
    "    ax[0].scatter(gRNAsbooCuts[0]-0.5,[-3]*len(gRNAsbooCuts[0]), marker='|', cmap='summer', c = 'red', s=Size, alpha = 1)\n",
    "\n",
    "    DFMutationsEach.iloc[len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd,:].plot(ax = ax[1], kind='bar', stacked=True, width = 1, color = sns.color_palette(\"Spectral\",12))\n",
    "    ax[1].set(title = Samples[Exp]+' n='+str(N[Exp])+'   '+Descriptions[Exp]+' total mismatches by type', ylim = [-5,Threshold],xlabel = 'position (bp)', ylabel = '% mismatch')\n",
    "    ax[1].set_xticklabels(GFPList, rotation = 0)\n",
    "    ax[1].grid(False)\n",
    "    ax[1].scatter(gRNAsbooFWD[0],[-3]*len(gRNAsbooFWD[0]), marker='4', cmap='summer', c = 'gray', s=Size, alpha = 0.75)\n",
    "    ax[1].scatter(gRNAsbooREV[0],[-3]*len(gRNAsbooREV[0]), marker='3', cmap='summer', c = 'gray', s=Size, alpha = 0.75)\n",
    "    ax[1].scatter(gRNAsbooCuts[0]-0.5,[-3]*len(gRNAsbooCuts[0]), marker='|', cmap='summer', c = 'red', s=Size, alpha = 1)\n",
    "    \n",
    "    DFMM = pd.DataFrame((MM-CtrMM)*100)\n",
    "    DFMM.iloc[len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd,:].plot(ax = ax[2], kind='bar', width = 1)\n",
    "    ax[2].set(title = Samples[Exp]+' n='+str(N[Exp])+'   '+Descriptions[Exp]+' total mismatches after background substraction', ylim = [-5,Threshold],xlabel = 'position (bp)', ylabel = '% mismatch')\n",
    "    ax[2].set_xticklabels(GFPList, rotation = 0)\n",
    "    ax[2].grid(False)\n",
    "    ax[2].scatter(gRNAsbooFWD[0],[-3]*len(gRNAsbooFWD[0]), marker='4', cmap='summer', c = 'gray', s=Size, alpha = 0.75)\n",
    "    ax[2].scatter(gRNAsbooREV[0],[-3]*len(gRNAsbooREV[0]), marker='3', cmap='summer', c = 'gray', s=Size, alpha = 0.75)\n",
    "    ax[2].scatter(gRNAsbooCuts[0]-0.5,[-3]*len(gRNAsbooCuts[0]), marker='|', cmap='summer', c = 'red', s=Size, alpha = 1)\n",
    "    \n",
    "    DFMutationsEach = (DFMutations[Exp]/N[Exp]-CtrMMeach)*100\n",
    "    SumMut = np.sum(DFMutationsEach,1)\n",
    "    AvMut = np.mean(SumMut.iloc[MutationsSelected])\n",
    "    DFMutationsEach.iloc[len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd,:].plot(ax = ax[3], kind='bar', stacked=True, width = 1, color = sns.color_palette(\"Spectral\",12))\n",
    "    ax[3].set(title = Samples[Exp]+' n='+str(N[Exp])+'   '+Descriptions[Exp]+' total mismatches by type after background substraction. '+str(round(AvMut))+'% av mut rate', ylim = [-5,Threshold],xlabel = 'position (bp)', ylabel = '% mismatch')\n",
    "    ax[3].set_xticklabels(GFPList, rotation = 0)\n",
    "    ax[3].grid(False)\n",
    "    ax[3].scatter(gRNAsbooFWD[0],[-3]*len(gRNAsbooFWD[0]), marker='4', cmap='summer', c = 'gray', s=Size, alpha = 0.75)\n",
    "    ax[3].scatter(gRNAsbooREV[0],[-3]*len(gRNAsbooREV[0]), marker='3', cmap='summer', c = 'gray', s=Size, alpha = 0.75)\n",
    "    ax[3].scatter(gRNAsbooCuts[0]-0.5,[-3]*len(gRNAsbooCuts[0]), marker='|', cmap='summer', c = 'red', s=Size, alpha = 1)\n",
    "    np.savetxt(DirOut+\"MismatchesEachType_\"+Samples[Exp]+'_'+Descriptions[Exp]+\".csv\", DFMutationsEach.iloc[len(PrimerFWD)+CutStart:len(GFPseqF)-len(PrimerREV)-CutEnd,:], delimiter=\",\")\n",
    "    np.savetxt(DirOut+\"MismatchesEachType_cols.csv\",GFPList, delimiter=\",\", fmt=\"%s\")\n",
    "    np.savetxt(DirOut+\"MismatchesEachType_rows.csv\",['AC','AT','AG','CA','CT','CG','TA','TC','TG','GA','GC','GT'] , delimiter=\",\", fmt=\"%s\")\n",
    "\n",
    "    f.tight_layout()\n",
    "    f.savefig(DirOut+\"allcombined_mismatched_samples_barplot_\"+Samples[Exp]+'_'+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.png\",dpi=300)\n",
    "    f.clf()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c46ba447-7aad-489b-8239-4b8bf2aaee3b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "180c3a94-d361-4588-ad19-5110bc77044b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# GET NUMBER PER READ\n",
    "\n",
    "num_cores = multiprocessing.cpu_count()\n",
    "pool_obj = multiprocessing.Pool()\n",
    "args = [(Samples,sample,DFall,REF,LibraryLength,MutationsSelected,GFPseqF,PrimerFWD,PrimerREV) for sample in Samples]\n",
    "[NumberMutations,Neach] = zip(*pool_obj.starmap(CountSelectedMismatches,args))\n",
    "print('done')\n",
    "\n",
    "# NumMut1 = NumberMutations[ControlSamples[WhichControl]]\n",
    "# NumMut2 = np.concatenate((NumberMutations[14:20]),axis = 0)\n",
    "# NumMut3 = np.concatenate((NumberMutations[24:28]),axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a18ae296-f5b8-4ec0-88aa-ffcb6b815227",
   "metadata": {},
   "outputs": [],
   "source": [
    "Control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d63e3c20-cd10-4339-9a15-ac9030532097",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#PLOT HISTOGRAMS # of SELECTED MUTATIONS\n",
    "#all histograms\n",
    "fig1, axes1 = plt.subplots(math.ceil(len(Samples)/6), 6, sharex=True, sharey=True, figsize=(20,12))\n",
    "#all histograms sustracting control\n",
    "fig2, axes2 = plt.subplots(math.ceil(len(Samples)/6), 6, sharex=False, sharey=False, figsize=(20,20))\n",
    "fig3, axes3 = plt.subplots(math.ceil(len(Samples)/6), 6, sharex=False, sharey=False, figsize=(20,12))\n",
    "DiffHistArray = np.zeros([len(Samples)-2,sum(MutationsSelected)+1])\n",
    "\n",
    "for Exp in range(0,len(Samples)-2):\n",
    "    for Control in ControlSamples:\n",
    "        if Exp >= Control:\n",
    "            #NumberMutationsCtr = NumberMutations[ControlSamples[Control]]\n",
    "            NumberMutationsCtr = NumberMutations[Control]\n",
    "    \n",
    "    plot = sns.histplot(NumberMutations[Exp],stat='percent',discrete=True,binwidth =1,color=\"blue\",alpha=0.5,ax=axes1[math.ceil((Exp+1)/6)-1,Exp-((math.ceil((Exp+1)/6))-1)*6],log_scale=(False, True))\n",
    "    plot = sns.histplot(NumberMutationsCtr,stat='percent',discrete=True,binwidth =1,color=\"orange\",alpha=0.5,ax=axes1[math.ceil((Exp+1)/6)-1,Exp-((math.ceil((Exp+1)/6))-1)*6],log_scale=(False, True))\n",
    "    plot.set(title=Samples[Exp]+' n='+str(N[Exp])+'\\n'+Descriptions[Exp],xlim = (-0.5,sum(MutationsSelected)),ylim=(0.5,100),ylabel = '% reads', xlabel = '# C-T / G-A mutations')\n",
    "    plot.set_yticks([0.5,1,10,25,50,90])\n",
    "    plot.set_yticklabels([0.5,1,10,25,50,90],fontsize = 9)\n",
    "\n",
    "    hist1, _ = np.histogram(NumberMutations[Exp],range=(0,sum(MutationsSelected)), bins = sum(MutationsSelected)+1)\n",
    "    hist2, _ = np.histogram(NumberMutationsCtr,range=(0,sum(MutationsSelected)), bins = sum(MutationsSelected)+1)\n",
    "    DiffHist = (hist1/sum(hist1) - hist2/sum(hist2))*100\n",
    "    DiffHist[DiffHist<0]=0\n",
    "    DiffHistArray[Exp,:] = DiffHist\n",
    "    CumDiffHist = np.cumsum(DiffHist)\n",
    "    TotalMut = np.sum(DiffHist)\n",
    "    \n",
    "    #print(CumDiffHist)\n",
    "    #DiffHist[0] = 100-sum(DiffHist)\n",
    "    plot = sns.barplot(x = list(range(0,sum(MutationsSelected)+1,1)), y = DiffHist,color=\"blue\",alpha=0.5, ax=axes2[math.ceil((Exp+1)/6)-1,Exp-((math.ceil((Exp+1)/6))-1)*6])\n",
    "    plot.grid(axis='x')\n",
    "    #plot.set(title=Samples[Exp]+' n='+str(N[Exp])+'\\n'+Descriptions[Exp],xlim = (-0.5,sum(MutationsSelected)),ylim=(0,10),yticks = range(0,11,2),xticks = range(0,sum(MutationsSelected),2), ylabel = '% reads', xlabel = '# C-T / G-A mutations - BG')   \n",
    "    plot.set(title=Samples[Exp]+' n='+str(N[Exp])+'\\n'+Descriptions[Exp]+' '+str(round(TotalMut))+'% mutated',xlim = (0,sum(MutationsSelected)),ylim=(0,10),yticks = range(0,16,3),xticks = range(0,sum(MutationsSelected),10), ylabel = '% reads', xlabel = '# C-T / G-A mutations - BG')   \n",
    "\n",
    "    plot = sns.lineplot(x = list(range(0,sum(MutationsSelected)+1,1)), y = CumDiffHist,color=\"orange\",alpha=0.5, ax=axes3[math.ceil((Exp+1)/6)-1,Exp-((math.ceil((Exp+1)/6))-1)*6])\n",
    "    plot = sns.barplot(x = list(range(0,sum(MutationsSelected)+1,1)), y = DiffHist,color=\"blue\",alpha=1, ax=axes3[math.ceil((Exp+1)/6)-1,Exp-((math.ceil((Exp+1)/6))-1)*6])\n",
    "    plot.grid(axis='x')\n",
    "    #plot.set(title=Samples[Exp]+' n='+str(N[Exp])+'\\n'+Descriptions[Exp],xlim = (-0.5,sum(MutationsSelected)),ylim=(0,10),yticks = range(0,11,2),xticks = range(0,sum(MutationsSelected),2), ylabel = '% reads', xlabel = '# C-T / G-A mutations - BG')   \n",
    "    plot.set(title=Samples[Exp]+' n='+str(N[Exp])+'\\n'+Descriptions[Exp]+' '+str(round(TotalMut))+'% mutated',xlim = (0,sum(MutationsSelected)),ylim=(0,100),yticks = range(0,50,10),xticks = range(0,sum(MutationsSelected),10), ylabel = '% reads', xlabel = '# C-T / G-A mutations - BG')   \n",
    "  \n",
    "    Exp+=1\n",
    "\n",
    "np.savetxt(DirOut+\"TableHistograms-BG.csv\", DiffHistArray, delimiter=\",\")\n",
    "# np.savetxt(DirOut+\"MismatchesEachType_cols.csv\",GFPList, delimiter=\",\", fmt=\"%s\")\n",
    "# np.savetxt(DirOut+\"MismatchesEachType_rows.csv\",['AC','AT','AG','CA','CT','CG','TA','TC','TG','GA','GC','GT'] , delimiter=\",\", fmt=\"%s\")\n",
    "    \n",
    "fig1.suptitle('% C-T and G-A mismatched reads to GFP in gRNAs before background substraction from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "fig1.tight_layout()\n",
    "fig1.savefig(DirOut+\"allcombined_mutations_histograms\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")\n",
    "fig1.clf()\n",
    "plt.close()\n",
    "\n",
    "fig2.suptitle('% C-T and G-A mismatched reads to GFP in gRNAs after background substraction from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "fig2.tight_layout()\n",
    "fig2.savefig(DirOut+\"allcombined_mutations_histograms_BG\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")\n",
    "fig2.clf()\n",
    "plt.close()\n",
    "\n",
    "fig3.suptitle('% C-T and G-A mismatched reads to GFP in gRNAs after background substraction from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "fig3.tight_layout()\n",
    "fig3.savefig(DirOut+\"allcombined_mutations_histograms_BG_cumulative\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")\n",
    "fig3.clf()\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a0efd3c-9d50-497b-94c3-d4b04355688f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9553b197-f671-4b71-96ec-57f03d77fc1b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# SELECT ALL BG MUTATIONS AND PLOT NUMBER PER READ\n",
    "# DFall = pd.read_csv(DirOut+\"DFall.csv\", sep=\",\")\n",
    "\n",
    "# args = [(Samples,sample,DFall,REF,LibraryLength,GFPseqF,PrimerFWD,PrimerREV) for sample in Samples]\n",
    "# [NumberMutationsAll,_] = zip(*pool_obj.starmap(CountAllMismatches,args))\n",
    "# print('done')\n",
    "\n",
    "# NumMutAll1 = NumberMutationsAll[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "021ca9e5-1834-4735-9525-1b37e5d05612",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#PLOT HISTOGRAMS # of ALL MUTATIONS\n",
    "# fig1, axes1 = plt.subplots(math.ceil(len(Samples)/6), 6, sharex=True, sharey=True, figsize=(20,10))\n",
    "\n",
    "# Exp = 0\n",
    "# for sample in Samples:\n",
    "#     if Exp in range(0,8):\n",
    "#         NumberMutationsAllCtr = NumMutAll1\n",
    "        \n",
    "#     plot = sns.histplot(NumberMutationsAll[Exp],stat='percent',discrete=True,binwidth =1,color=\"blue\",alpha=0.5,ax=axes1[math.ceil((Exp+1)/6)-1,Exp-((math.ceil((Exp+1)/6))-1)*6],log_scale=(False, True))\n",
    "#     plot = sns.histplot(NumberMutationsAllCtr,stat='percent',discrete=True,binwidth =1,color=\"orange\",alpha=0.5,ax=axes1[math.ceil((Exp+1)/6)-1,Exp-((math.ceil((Exp+1)/6))-1)*6],log_scale=(False, True))\n",
    "#     plot.set(title=Samples[Exp]+' n='+str(N[Exp])+'\\n'+Descriptions[Exp],xlim = (0,20), ylim=(0.5,100),ylabel = '% reads', xlabel = '# C-T / G-A mutations')\n",
    "#     plot.set_yticks([0.5,1,10,25,50,90])\n",
    "#     plot.set_yticklabels([0.5,1,10,25,50,90],fontsize = 9)\n",
    "#     Exp+=1\n",
    "    \n",
    "# fig1.suptitle('% mismatched reads to GFP before background substraction from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "# fig1.tight_layout()\n",
    "# fig1.savefig(DirOut+\"allBGcombined_mutations_histograms\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")\n",
    "# fig1.clf()\n",
    "# plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac373348-1960-4992-92c5-43bfbd73b277",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CALCULATE AVERAGE MUTATION RATES OVER gRNAs\n",
    "fig1, ax = plt.subplots(1, 2, sharex=False, sharey=True, figsize=(18,5))\n",
    "\n",
    "samplei = 0\n",
    "# for sample in Samples[0:-2]:\n",
    "#AIDExperiments = [3, 4, 5, 11, 12, 13, 21, 22, 23]\n",
    "#AIDExperiments = range(0,len(Samples)-2)\n",
    "AIDExperiments = ExperimentSamples\n",
    "for sample in AIDExperiments:\n",
    "    AllMismatchesEachSample = MatrixMM_BG[samplei,:]\n",
    "    EachMismatchesEachSample = DFMutations_BG[samplei]\n",
    "    CTGAMismatchesEachSample = EachMismatchesEachSample['CT']+EachMismatchesEachSample['GA']\n",
    "    MaxgRNAlength = max(gRNAsDF['End']-gRNAsDF['Start']+1)\n",
    "    MgRNAsMMAlignedStart = np.zeros((len(gRNAsDF),MaxgRNAlength),float)\n",
    "    MgRNAsMMAlignedPAM = np.zeros((len(gRNAsDF),MaxgRNAlength),float)\n",
    "\n",
    "#     MgRNAsCTGA = np.zeros((len(gRNAsDF),MaxgRNAlength),float)\n",
    "    for index, row in gRNAsDF.iterrows():  \n",
    "        Start = row['Start']+len(PrimerFWD)\n",
    "        End = row['End']+len(PrimerFWD)\n",
    "        Orientation = row['Orientation']\n",
    "\n",
    "        if Orientation == '+':\n",
    "            isC = 'C'== np.array(list(GFPseqF[Start-1:End]))\n",
    "            AvAllgRNAs = AllMismatchesEachSample[Start-1:End]*isC\n",
    "            AvAllgRNAs[AvAllgRNAs==0] = None\n",
    "#             AvCTGAgRNAs = CTGAMismatchesEachSample[row['Start']-1:row['End']]\n",
    "        else:\n",
    "            isC = 'G'== np.flip(np.array(list(GFPseqF[Start-1:End])))\n",
    "            AvAllgRNAs = np.flip(AllMismatchesEachSample[Start-1:End])*isC\n",
    "            AvAllgRNAs[AvAllgRNAs==0] = None\n",
    "#             AvCTGAgRNAs = np.flip(CTGAMismatchesEachSample[row['Start']-1:row['End']])\n",
    "        if len(AvAllgRNAs) < MaxgRNAlength:\n",
    "            #aligning by PAM\n",
    "            AvAllgRNAsEnd = [None]*(MaxgRNAlength-len(AvAllgRNAs)) + list(AvAllgRNAs)\n",
    "#             AvCTGAgRNAs = [None]*(MaxgRNAlength-len(AvCTGAgRNAs)) + list(AvCTGAgRNAs)\n",
    "            #aligning by start gRNA\n",
    "            AvAllgRNAsStart = list(AvAllgRNAs) + [None]*(MaxgRNAlength-len(AvAllgRNAs))\n",
    "#             AvCTGAgRNAs = list(AvCTGAgRNAs) + [None]*(MaxgRNAlength-len(AvCTGAgRNAs))\n",
    "            #normalize to gRNA length??\n",
    "        else:\n",
    "            AvAllgRNAsStart = AvAllgRNAs\n",
    "            AvAllgRNAsEnd = AvAllgRNAs\n",
    "        MgRNAsMMAlignedStart[index,:] = AvAllgRNAsStart\n",
    "        MgRNAsMMAlignedPAM[index,:] = AvAllgRNAsEnd\n",
    "    sns.lineplot(x=range(1,MaxgRNAlength+1,1),y=np.nanmean(MgRNAsMMAlignedStart,0)*100,marker = 'o', ax = ax[0])\n",
    "    sns.lineplot(x=range(-MaxgRNAlength,0,1),y=np.nanmean(MgRNAsMMAlignedPAM,0)*100, marker = 'o',ax = ax[1],palette='crest')\n",
    "    samplei += 1\n",
    "ax[0].set(title='Probablility of mutating a C at a given gRNA position',ylabel='Average % mutation of all gRNAs', xlabel='gRNA position (bp)')\n",
    "ax[0].legend(labels=list(Descriptions[AIDExperiments]))\n",
    "ax[1].set(title='Probablility of mutating a C at a given gRNA position',ylabel='Average % mutation of all gRNAs', xlabel='Distance from PAM (bp)')\n",
    "sns.set_palette('Paired')\n",
    "# ax[0].set(legend = Samples(ExperimentSamples))\n",
    "ax[1].legend(labels=list(Descriptions[AIDExperiments]))\n",
    "fig1.savefig(DirOut+\"mutation_profile_avgRNAs\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")\n",
    "#fig1.savefig('../'+\"mutation_profile_avgRNAs\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c91a2359-f42d-4a18-8d57-2663918b62a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "# obj0, obj1, obj2 are created here...\n",
    "\n",
    "# Saving the objects:\n",
    "with open(Path+'plots-sup/data.pkl', 'wb') as f:  # Python 3: open(..., 'wb')\n",
    "    pickle.dump([Mismatch, DFMutations, Indels, N, Samples, Descriptions, ControlSamples, ExperimentSamples, NumberMutations, Neach, MgRNAsMMAlignedPAM], f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f0ff267-bc8b-4916-8186-60611d77302e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "# Getting back the objects:\n",
    "with open(Path+'plots-sup/data.pkl','rb') as f:  # Python 3: open(..., 'rb')\n",
    "    Mismatch, DFMutations, Indels, N, Samples, Descriptions, ControlSamples, ExperimentSamples, NumberMutations, Neach,MgRNAsMMAlignedPAM = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b55d39f-e9db-41c6-9dc4-09f5a2575eaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SubDF = DFall[DFall['Sample'] == '#11']\n",
    "# Cbar_kws={'ticks':[],'label':'N      A      C      T      G      D','boundaries':[0,1,2,3,4,5,6]}\n",
    "# hmA = sns.clustermap(SubDF.iloc[0:,len(PrimerFWD)+15:len(GFPseqF)-len(PrimerREV)-75],cmap=\"viridis\",vmin=0, vmax=5,figsize=(16, 16),cbar_pos=(0.1, 0.5, 0.05, 0.12),col_cluster=False,row_cluster=True,cbar_kws=Cbar_kws)\n",
    "# #hmM = sns.clustermap(SubDF.iloc[:,0:LibraryLength],col_cluster=False,row_cluster=False,figsize=(10, 10),vmin=0,vmax = Threshold, yticklabels=Descriptions, cbar_kws={'label': '% mismatch'})\n",
    "# hmA.ax_cbar.set_position((0.075,0.6, .03, .2))\n",
    "# hmA.fig.suptitle('% mismatched reads to GFP in each position after background subtraction from '+ str(round(len(FilesIn)/len(Files)*100)) + \"% data (\"+ str(len(FilesIn))+\"/\"+str(len(Files))+\" files)\")\n",
    "# hmA.savefig(DirOut+\"#11_heatmapAligned_SortedIndex_\"+str(len(FilesIn))+\"-\"+str(len(Files))+\"files.png\")\n",
    "# hmM.fig.clf()\n",
    "# plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8432e655-0426-4108-9078-961f1b171a7b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
